### PR TITLE
Issue/75 extractkb

### PR DESCRIPTION
**Pull request by vitalych**
https://github.com/S2E/s2e-env/issues/75

### CODE REVIEW

File: windows/scripts/extract_kernels.py [hunk 1]
```diff
@@ -56,6 +56,9 @@
 
 NT_PATTERN = re.compile('(ntoskrnl|ntkrnlmp|ntkrnlpa)$')
 
+# This will work only on msys
+EXPAND_PATH = '/c/Windows/System32/expand.exe'
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py]: Should use `os.path.join`

File: windows/scripts/extract_kernels.py [hunk 2]
```diff
@@ -124,7 +99,6 @@ def cleanup(self, warn=False):
             if warn:
                 print('Implicitly cleaning up {!r}'.format(self))
 
-
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py:127]: x2 blank lines between Python functions (except if they are class methods, in which case it's only a single line).

File: windows/scripts/extract_kernels.py [hunk 3]
```diff
@@ -177,23 +150,147 @@ def seven_zip_extract(source, wildcard_includes, wildcard_excludes=None,
...
+
+def extract_kernels_wim(output_dir, iso_file):
+    iso_basename, _ = os.path.splitext(os.path.basename(iso_file))
+
+    with TemporaryDirectory() as temp_dir:
+        seven_zip_extract(iso_file, ['sources/install.wim'], dest_dir=temp_dir)
+        wim = os.path.join(temp_dir, 'install.wim')
+
+        # Rename the wim to have more informative kernel name
+        dest_wim = os.path.join(temp_dir, iso_basename + '.wim')
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py]: `'%s.wim' % iso_basename`

File: windows/scripts/extract_kernels.py [hunk 4]
```diff
@@ -159,6 +162,20 @@ def seven_zip_extract(source, wildcard_includes, wildcard_excludes=None,
     if proc.returncode:
         raise Exception(u'Failed to extract file %s from %s: "%s"' % (wildcard_includes, source, stderr))
 
+# We can't use 7z to extract files from CABs because some files inside may use delta encoding.
+# Such a cab file would contain XML files describing the various chunks of a file.
+# 7z would get us the chunks and we need the reassembled files.
+def cab_extract(source, dest_dir, pattern = '*'):
+    args = [EXPAND_PATH, source, '-F:%s' % (pattern), dest_dir]
+
+    print("    %s - %s" % (" ".join(args), dest_dir))
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py]: Single quotes

File: windows/scripts/extract_kernels.py [hunk 5]
```diff
@@ -241,13 +265,38 @@ def is_valid_kernel(path):
...
+            ret.append(filepath)
+
+    return ret
+
 def extract_kernels_from_container(output_dir, container, files):
+    container_is_dir = False
+
+    if os.path.isdir(container):
+        if files:
+            raise Exception('Files must be null if container is a folder')
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py]: What does "null" mean here? Null is not really a thing in Python - do you mean `None`? Or an empty list?

File: windows/scripts/extract_kernels.py [hunk 6]
```diff
@@ -280,12 +329,31 @@ def extract_kernels_from_iso(output_dir, iso_file):
     else:
         print("Invalid install iso")
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py]: Use single quotes

File: windows/scripts/gendriver.sh [hunk 7]
```diff
@@ -38,9 +40,29 @@ if [ ! -d $ISO_DIR ]; then
...
+
+# Download patches for driver signature checks
+download_patch https://download.microsoft.com/download/C/8/7/C87AE67E-A228-48FB-8F02-B2A9A1238099/Windows6.1-KB3033929-x64.msu
+download_patch https://download.microsoft.com/download/3/7/4/37473F39-5728-4153-9A25-64C09DE9ED52/Windows6.1-KB3033929-x86.msu
+
+# Extract kernels from KBs first
+./scripts/extract_kernels.py  --iso-dir "$KB_DIR" -o "$OUTPUT_DIR"
+
+# Extract ISOs
+# ./scripts/extract_kernels.py  --iso-dir "$ISO_DIR" -o "$OUTPUT_DIR"
```
Comment (adrianherrera) [inline windows/scripts/gendriver.sh]: Why is this line still here? Should it be removed?

File: windows/scripts/extract_kernels.py [hunk 8]
```diff
@@ -53,33 +54,7 @@
...
-# We use the wildcard pattern "nt*.exe" to find kernel executables in a Windows
-# ISO. Unforuntaly, this can also match other executables that are not related
-# to the kernel. A blacklist of these executables is maintained here.
-NT_BLACKLIST = [
-    'ntprint.exe',
-    'ntbackup.exe',
-    'ntvdm.exe',
-]
-
+NT_PATTERN = re.compile('(ntoskrnl|ntkrnlmp|ntkrnlpa)$')
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py]: Even though you don't have to escape any special character, it's still good practice to use `r'(ntoskrnl|ntkrnlmp|ntkrnlpa)$'` for regex strings

File: windows/scripts/extract_kernels.py [hunk 9]
```diff
@@ -177,23 +150,147 @@ def seven_zip_extract(source, wildcard_includes, wildcard_excludes=None,
...
+
     # Run 7-Zip
-    proc = subprocess.Popen(args, stdout=DEVNULL, stderr=subprocess.PIPE,
-                            cwd=dest)
+    proc = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=dest_dir)
     _, stderr = proc.communicate()
 
-    return proc.returncode, stderr
+    if proc.returncode:
+        raise Exception(u'Failed to extract file %s from %s: "%s"' % (wildcard_includes, source, stderr))
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py]: Maybe prepend string with `[\u2717]`, like the other error messages do

File: windows/scripts/extract_kernels.py [hunk 10]
```diff
@@ -177,23 +150,147 @@ def seven_zip_extract(source, wildcard_includes, wildcard_excludes=None,
...
+def seven_zip_list(source):
+    args = ['7z', 'l', source]
+    proc = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = proc.communicate()
+
+    files = []
 
+    for line in stdout.splitlines():
+        els = line.split()
+        if len(els) == 6:
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py]: Can you maybe add a small comment just to explain the significance of these magic numbers? That way if the 7zip output format ever changes it's easier to update the script. Is `els[5]` a file name? A file path? It's not immediately clear what this function returns

File: windows/scripts/extract_kernels.py (hunk 8 above)
Comment (adrianherrera) [reply windows/scripts/extract_kernels.py]: Does it make more sense to call this constant `KERNEL_PATTERN` or something like that?

File: windows/scripts/extract_kernels.py [hunk 11]
```diff
@@ -177,23 +150,147 @@ def seven_zip_extract(source, wildcard_includes, wildcard_excludes=None,
 
     args.append(source)
 
+    print("    %s - %s" % (" ".join(args), dest_dir))
```
Comment (adrianherrera) [inline windows/scripts/extract_kernels.py]: single quotes

**Review by adrianherrera — COMMENTED** [review]
Run `run_pylint.py` script to check if it conforms to our/PEP8 style.

File: windows/scripts/gendriver.sh (hunk 7 above)
Comment (vitalych) [reply windows/scripts/gendriver.sh]: Forgot to uncomment after debugging

**Comment by vitalych** [comment]
@adrianherrera I addressed your comments

**Review by adrianherrera — APPROVED** [review]

**Pull request merged**